src/recode.jl

```diff
@@ -96,34 +101,29 @@ function recode!(dest::AbstractArray{T}, src::AbstractArray, default::Any, pairs
     dest
 end
 
-function recode!(dest::CategoricalArray{T}, src::AbstractArray, default::Any, pairs::Pair...) where {T}
-    if length(dest) != length(src)
-        throw(DimensionMismatch("dest and src must be of the same length (got $(length(dest)) and $(length(src)))"))
-    end
-
-    opt_pairs = map(optimize_pair, pairs)
+function _recode!(dest::CategoricalArray{T, <:Any, R}, src::AbstractArray, default::Any, pairs) where {T, R}
+    recode_to = last.(pairs)
+    recode_from = first.(pairs)
 
-    vals = T[p.second for p in opt_pairs]
-    default !== nothing && push!(vals, default)
+    vals = convert.(T, recode_to)
+    vals = default === nothing ? vals : (vals..., default)
 
     levels!(dest.pool, filter!(!ismissing, unique(vals)))
     # In the absence of duplicated recoded values, we do not need to lookup the reference
     # for each pair in the loop, which is more efficient (with loop unswitching)
     dupvals = length(vals) != length(levels(dest.pool))
 
     drefs = dest.refs
-    pairmap = [ismissing(v) ? 0 : get(dest.pool, v) for v in vals]
-    defaultref = default === nothing || ismissing(default) ? 0 : get(dest.pool, default)
+    pairmap = [ismissing(v) ? zero(R) : get(dest.pool, v) for v in vals]
+    defaultref = default === nothing ? nothing : ismissing(default) ? 0 : get(dest.pool, default)
+
     @inbounds for i in eachindex(drefs, src)
         x = src[i]
 
-        for j in 1:length(opt_pairs)
-            p = opt_pairs[j]
-            # we use isequal and recode_in because we cannot really distinguish scalars from collections
-            if x ≅ p.first || recode_in(x, p.first)
-                drefs[i] = dupvals ? pairmap[j] : j
-                @goto nextitem
-            end
+        j = Compat.@inline findfirst(y -> isequal(x, y) || recode_in(x,y), recode_from)
```

Comment:
```suggestion
        # we use isequal and recode_in because we cannot really
        # distinguish scalars from collections
        j = Compat.@inline findfirst(y -> isequal(x, y) || recode_in(x, y), recode_from)
```